src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt

```diff
@@ -0,0 +1,279 @@
+        }
+        for (clause in solver.constraints) {
+            clause.forEachIndexed { ind, lit ->
+                if (lit > 0) {
+                    clause[ind] = newNumeration[lit]
+                } else {
+                    clause[ind] = -newNumeration[-lit]
+                }
+            }
+        }
+        solver.varsNumber = newSize
+        countOccurrence()
+        updateSig()
+    }
+
+
+    private fun addResolvents(ind: Int) {
+        for (cl1 in litOccurrence[litPos(ind)]) {
+            if (isClauseDeleted[cl1]) {
+                continue
+            }
+            for (cl2 in litOccurrence[litPos(-ind)]) {
+                if (isClauseDeleted[cl2]) {
+                    continue
+                }
+                val newClause = solver.constraints[cl1].toMutableSet()
+                newClause.remove(ind)
+                // check if clause is tautology
+                if (solver.constraints[cl2].any { newClause.contains(-it) }) {
+                    continue
+                }
+                newClause.addAll(solver.constraints[cl2])
+                newClause.remove(-ind)
+                solver.constraints.add(Clause(newClause.toMutableList()))
+                isClauseDeleted.add(false)
+                for (lit in newClause) {
+                    litOccurrence[litPos(lit)].add(solver.constraints.lastIndex)
+                }
+            }
+        }
+        for (clause in litOccurrence[litPos(ind)]) {
+            isClauseDeleted[clause] = true
+        }
+        for (clause in litOccurrence[litPos(-ind)]) {
+            isClauseDeleted[clause] = true
+        }
+    }
+
+    // remove subsumed clauses
+    private fun removeSubsumedClauses() {
+        val uselessClauses = mutableSetOf<Int>()
+        val duplicateClauses = mutableSetOf<Int>()
+        val markedClauses = MutableList(solver.constraints.size) { false }
+
+        // going from the end because smaller clauses appear after big one
+        for (ind in solver.constraints.lastIndex downTo 0) {
+            if (!markedClauses[ind]) {
+                findSubsumed(ind).forEach {
+                    if (!markedClauses[it]) {
+                        if (ind.clauseSize() < it.clauseSize()) {
+                            markedClauses[it] = true
+                            uselessClauses.add(it)
+                        } else if (ind.clauseSize() == it.clauseSize()) {
+                            duplicateClauses.add(it)
+                        }
+                    }
+                }
+            }
+        }
+
+        val copiedClauses = duplicateClauses.map { solver.constraints[it] }
+
+        solver.constraints.removeAll(uselessClauses.map { solver.constraints[it] })
+        // remove duplicate clauses and leave 1 copy of each
+        solver.constraints.removeAll(copiedClauses)
+        solver.constraints.addAll(copiedClauses)
+        countOccurrence()
+        updateSig()
+    }
+
+    private fun removeTautologies() {
```

Comment:
What is a tautology?

If you are trying to remove the duplicate literals, then (1) consider doing this in `Solver::addClause` method, since this is not a responsibility of a `Preprocessor`, given a fact that the literals can't just duplicate during the solving process, (2) consider just using `toSet().toList()` or even `.distinct()` methods.